### ISSUE TITLE
JSON Schema for the OpenAPI 3.0 Specification

### ISSUE THREAD

**Issue by tedepstein** [OAI/OpenAPI-Specification#1032]
It looks like we don't yet have a JSON Schema to describe the OpenAPI 3.0 spec.  The [schemas folder](https://github.com/OAI/OpenAPI-Specification/tree/master/schemas) has 1.2 and 2.0 subfolders, but nothing for 3.0 yet.

We'd like to contribute to this.  Is anyone working on a JSON Schema for the 3.0 spec yet?

**Comment by MikeRalphson** [comment]
@timburks and I are [working on one](https://github.com/googleapis/gnostic/blob/master/OpenAPIv3/openapi-3.0.json) over at Google [Gnostic](https://github.com/googleapis/gnostic), which is generated from the spec itself.

My latest version is [here](https://github.com/Mermade/swagger2openapi/blob/master/schemas/openapi-3.json), which incorporates a few manual fixes. Help improving and testing it would be gratefully received.

**Comment by tedepstein** [comment]
@MikeRalphson , @timburks, great idea to auto-generate a schema from the written spec. I haven't looked at it carefully yet, but at first glance it seems like at least a good starting point.

@earth2marsh, you mentioned in your [Google Cloud Next talk](https://www.youtube.com/watch?v=4lBMQteMd6Y) that schema development was one of the next items on your agenda.  Does the TDC or the Apigee team have an OpenAPI v3 JSON schema in some stage of development...?

**Comment by brandur** [comment]
I hope this doesn't seem too much like a content-free +1, but I'd also love to see a JSON schema along with the specification.

I don't really trust myself to be able to accurately read the spec in its entirely, so for 2.0 I fell back heavily on using the included schemas to verify that what I'm generating is actually intelligible (and it worked, they caught many problems).

**Comment by timburks** [comment]
My latest draft is [here](https://github.com/googleapis/gnostic/blob/openapi-v3.0.0-rc2/OpenAPIv3/openapi-3.0.json), semi-automatically generated from the 3.0.md text (v3.0.0-rc2).

**Comment by brandur** [comment]
@timburks @MikeRalphson That's awesome. Thanks for the continued work on that! Just out of curiosity because I don't know the project dynamics here, is there any chance that your JSON schema can be sent upstream and be officially supported as part of the 3.0 release? I haven't seen or heard much from the project maintainers, but I figure that you may have spoken to them through another channel.

**Comment by timburks** [comment]
@brandur To keep it regular, I'd like to avoid hand-edits to the schema as much as possible, so I haven't submitted it to the specification repo yet. I expect to do that after we have a final spec.

**Comment by timburks** [comment]
It seemed good to get this into discussion, so I've posted a [PR](https://github.com/OAI/OpenAPI-Specification/pull/1236) containing our latest draft.

**Comment by mrglavas** [comment]
This would be very beneficial for library / tools developers that could use the JSON schema directly for (partial) validation and for generation of an OpenAPI v3.0 model in the language of their choice (not necessarily Java). Could also see this used for generating sample OpenAPI documents from the schema to make it easier to test that libraries and client applications are working correctly.

**Comment by sandorfr** [comment]
yes this is much needed  thanks @timburks for the draft. seems that you have a more recent version here now https://github.com/Mermade/swagger2openapi/blob/master/schemas/openapi-3.0.json ?

**Comment by MikeRalphson** [comment]
@sandorfr that schema is based on @webron's work in this PR #1270 - my copy you linked to has the object descriptions ported over from @timburks schema.

**Comment by MrBlenny** [comment]
Looks like I missed the boat here. There are typescript interfaces here: https://github.com/metadevpro/openapi3-ts

These are not fully annotated so they cannot really be considered complete. Going from typescript -> json using something like [typescript-json-schema](https://github.com/YousefED/typescript-json-schema) really improves readability.

**Comment by mohawk2** [comment]
Will this issue manage to be open a full year after the alleged release of OpenAPI 3? :-)

**Comment by mohawk2** [comment]
To follow up to my comment there: people, without a schema, you do not have a spec, you have an aspiration. Please pick one of the two schemas (probably the generated one from @timburks ) and declare it official. If either the spec or the schema need updating, you can just release minor version-updates.

No apologies for mass-@-ing here: @darrelmiller @earth2marsh @MikeRalphson @webron @whitlockjc - please execute something on this.

Release early, release often! One out of two won't be bad :-)

Cc the Perl module issue: https://github.com/jhthorsen/json-validator/issues/41

**Comment by darrelmiller** [comment]
@mohawk2 When we do release a final version of JSON Schema, please do not use JSON Schema to guarantee an OpenAPI document is valid.  It cannot do that.  There are numerous constraints in the written specification that cannot be expressed in JSON Schema.

One of the reasons I initially pushed back on the creation of a JSON Schema for V3 is because I feared that people would try to use it as a document validator.  However, I was convinced by other TSC members that there were valid uses of a schema beyond validation.

**Comment by jwalton** [comment]
@jberger I think what @handrews and @darrelmiller are getting at, to give a concrete example, is something like `operationId` in the [operation object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#operation-object).  JSON Schema can validate that the operationId field is either not present, or if it is present then it is a string (this is the "structural" part).  But JSON Schema cannot validate that `operationId` for a given operation object is unique compared to all other operation objects in the OpenAPI document (this is the "semantic" part).  JSON Schema doesn't "understand" a document, it just checks to see if required fields are there and the fields are the right type.

If you look at swagger-editor, it does [structural validation](https://github.com/swagger-api/swagger-editor/blob/12a9c3b6ec9bc22157cb921994bb98986c97a291/src/plugins/validate-json-schema/structural-validation/oas3-schema.js) using JSON Schema.  But then it also has a bunch of code for [semantic validation](https://github.com/swagger-api/swagger-editor/tree/a930a70c9999a369e4abcaa8265bfa66e4aa07d6/test/plugins/validate-semantic) outside of that.

I'm debating writing a bespoke validator for OpenAPI 3.0,, actually, as a minor typo in swagger-editor will often result in four or five errors (usually the JSON Schema validation will work out something is wrong in a given object, but then will also mark the parent object as not conforming, and the parent of that, and you get a lot of "cruft" errors that aren't really related to your actual problem).

**Comment by darrelmiller** [comment]
@jberger I chose to write a bespoke validator in .NET for V3.  Using JSON Schema and some additional validation code is definitely a viable approach.  Maybe one day, JSON Schema would be able to express all the constraints in the OpenAPI spec, but I suspect some are going to be really hard.

Writing a validator is a fair amount of work.  But hopefully if it done well it only needs to be done once per platform.  I chose to do a custom one because as @jwalton mentioned, the errors that you get from JSON schema can sometimes be very confusing.  I wanted to be able to generate errors that could easily be understood to speed up debugging time.

Having said this, the validation part has actually been easier than dealing with V2 to V3 conversion and the hardest part $ref resolution.

One question I have about using JSON Schema for validation:  How would you handle YAML documents?  Do you do a YAML to JSON conversion and then use a JSON schema validator?  That seems like it would be fairly computationally expensive.

**Comment by sandorfr** [comment]
@darrelmiller my use of the schema would have nothing to do with validation but generating typescript definition and having more accurate code editor completion/suggestions.

**Comment by brandur** [comment]
I want to come in with a counterpoint here because based on emoji reactions on comments alone, I have a feeling that I'm part of a not-very-vocal majority. Most of us who are interested in a JSON schema for OpenAPI 3 are aware that it doesn't _guarantee_ perfect validity, but we're interested in a formally maintained JSON schema nonetheless.

As an example, we're generating API descriptions like OpenAPI 3 using a generator that consumes an internal DSL. Given that all this code has one user (us) and is extremely custom, it's easy to make mistakes during implementation, and also easy to introduce regressions as things change. Internalizing a JSON schema and validating against it gets us 99% of the way to a valid spec, and we run it as part of every build in our CI suite.

Writing a special validator in every possible language would get you better errors, but it strikes me as quite impractical — writing new software is time consuming, and maintaining software is even more time consuming. It's also by no means guaranteed — you can easily find dozens (if not hundreds) of Swagger or OpenAPI-related packages on GitHub that have fallen into disrepair since the original good intentions of their authors. These JSON schemas are already most of the way there, and provide a language-agnostic solution since pretty much all mainstream languages can be expected to have a JSON schema validator written for them.

I'd really implore the maintainers here to adopt one of these JSON schemas. It might not be perfect, but it'll be a good enough solution for the vast majority of us, and infinitely better than the situation on the ground today.

> One question I have about using JSON Schema for validation: How would you handle YAML documents? Do you do a YAML to JSON conversion and then use a JSON schema validator? That seems like it would be fairly computationally expensive.

Many validators don't just operate directly on JSON and have an option to validate a generic (language-specific) hash structure (e.g. `Hash` in Ruby or `map[string]interface{}` in Go). As one data point, we generate an in-memory structure and dump to both JSON and YAML. Even if that wasn't the case though, moving between JSON and YAML is practically speaking, a fast operation.

**Comment by mohawk2** [comment]
Jeepers kitty-hugging Cripes. Nobody expects the v3 schema to be more of a perfect guarantee than the v2 schema (as said above, consider `operationId`). Just release a schema and let the dice fall where they may.

**Comment by mohawk2** [comment]
Otherwise, what is the actual point of OpenAPI? Just declare failure and get on with the rest of your lives.

**Comment by sandorfr** [comment]
@mohawk2 what did you do ! It is important to remain respectful even when through disagreement in the open source community.

That said, I hate when one developer don't want to implement a wanted and useful feature because some other developer might "use it wrong" according to their own "dogma".

**Comment by mohawk2** [comment]
@sandorfr Not trying to create conflict. Instead trying to break a year-long (and counting) logjam.

What are you trying to do? I fear trying to create the appearance of spurious unity?

**Comment by darrelmiller** [comment]
@sandorfr What is the reason for doing typescript generation from the OAS schema?  I can understand doing typescript generation from the Schemas in an OpenAPI Description, but creating typescript for the OAS Schema seems like a one off thing.

**Comment by jwalton** [comment]
@mohawk2 "Otherwise, what is the actual point of OpenAPI? Just declare failure and get on with the rest of your lives." is a bit over the top, don't you think?  There are plenty of examples of standards with no ABNF/ASN.1/schema-of-the-month-here, or where the ABNF is present but non-normative.  I'd like to see an "official" JSON Schema, too, but let's not get carried away here.

**Comment by darrelmiller** [comment]
@mohawk2 You have an unusual technique for motivating volunteers to do more free work!  I do understand your frustration, but all of us have limited free time and lots of work that needs to be done.  

It is my understanding that the Schema in PR #1270 is mostly complete.  If the primary goal is to get most of the way there, then it would great if folks started trying out that schema in the tools they are using and provide feedback on what works and what doesn't work.  That way folks are unblocked and their feedback can help get us to the place we all want to be.

If folks don't want to help getting us to a final version of the schema, then they will need to be patient until completing the OAS schema rises to the top of the priority list.

**Comment by darrelmiller** [comment]
@brandur  It is true that we all come with different contexts and different requirements.  For my company we decided that a from scratch validator was worth the effort because there could be literally thousands of developers verifying many hundreds of specs using the library.

I understand that for some people the JSON Schema will be sufficient.  However, for every OpenAPI user that is aware of the limitations of JSON Schema, there are 100 that aren't. And when you are on the front line of answering GitHub issues, you tend to be extra enthusiastic about building a pit of success.

As I mentioned in my last comment, I would encourage you to try and use the schema that is in PR #270 and provide us feedback so that we can get it finished ASAP.

**Comment by jberger** [comment]
Thank you all for the well reasoned replies. I especially echo the arguments of @brandur that having the leg up of starting with an openapi spec gets us so much closer to a working validator (as is true of the v2 implementation in Perl).

That said I feel I should apologize on behalf of our project. Mohawk is a sometime contributor to the Perl JSON::Validator project and thus he came here from discussions we've had in which we've been asked to implement v3 by users but are waiting on a spec. I wish he hadn't brought the harsh tone to this otherwise constructive discussion. As open source developers ourselves we know the feeling of being spoken to unkindly about work we do of the goodness of our hearts and the deflating and demotivating effect that cam have. My apologies.

**Comment by sandorfr** [comment]
@darrelmiller I develop some tooling around OAS in node/typescript, having a good typing helps me doing so.

**Comment by KES777** [comment]
@darrelmiller Hi.
>If folks don't want to help getting us to a final version of the schema
I want yesterday to try use OpenAPI v3 schema for my project. 
How can I tell feedback if there is no schema I can try?
There is no [v3 schema](https://github.com/OAI/OpenAPI-Specification/tree/master/schemas). There is nothing to try.
So we just decide to use v2...

IMHO: You may release draft1 and draft2 proposed schemas at this channel. Describe what is not finished. And you get feedback much sooner.

**Comment by MikeRalphson** [comment]
@KES777 the two draft schemas are not hard to find from the open PRs (#1236 and #1270). They are [here](https://github.com/timburks/OpenAPI-Specification/blob/5f3a9a7f547785c3f38001fc0ef82ff621d1bb79/schemas/v3.0/schema.json) and [here](https://github.com/OAI/OpenAPI-Specification/blob/49e784d7b7800da8732103aa3ac56bc7ccde5cfb/schemas/v3.0/schema.yaml) respectively.

**Comment by KES777** [comment]
@MikeRalphson  That took a couple of hours. It would be very easily if they were [here](https://github.com/OAI/OpenAPI-Specification/tree/master/schemas)

**Comment by MikeRalphson** [comment]
Apologies for my tardiness. The schemas have been linked to previously in this thread. If the draft schemas were published where you suggest, it would be hard to ensure people treated them **as drafts**.

**Comment by KES777** [comment]
filename: `schema-3.0-draft1.json`, `schema-3.0-draft2.json`

**Comment by MikeRalphson** [comment]
Please consider whether it is likely such things have already been thought about? The two schemas are alternatives, not different drafts of the same schema, which might be implied by the naming convention you suggest.

**Comment by KES777** [comment]
name those as `schema-3.0-alternative`, but publish them more frequently as @mohawk2 noticed. Currently they are almost lost in discussions... (

**Comment by MikeRalphson** [comment]
[IRTHGTTRIGSMA](https://acronyms.thefreedictionary.com/IRTHGTTRIGSMA)

**Comment by shadowcat-mst** [comment]
@mohawk2 

Given @darrelmiller's comment of:

> As I mentioned in my last comment, I would encourage you to try and use the schema that is in PR #270 and provide us feedback so that we can get it finished ASAP.

I look forward to listening to you rant on IRC in the process of getting to be able to provide feedback - well volunteered!

**Comment by darrelmiller** [comment]
Gotta love the passion and enthusiasm for getting this issue resolved!

![](https://media.giphy.com/media/1GJRIgTY4sS6k/giphy.gif)

**Comment by mohawk2** [comment]
@darrelmiller It would help reduce confusion if you could edit the comment where you linked to #270 to link to what I think you meant, which was #1270 :-)

**Comment by darrelmiller** [comment]
@mohawk2 That definitely would help :-)

**Comment by mohawk2** [comment]
@darrelmiller I had a look in the repo and didn't see anything like this, so it might well be considered far out of scope. However, do you think there is any value in a corpus of deliberately-structurally-invalid OAS specs for v3 and/or v2 that could be used to exercise at least a structural validator (most useful here given there isn't yet a final v3 schema), and perhaps additionally a separate corpus of semantically-invalid ones?

**Comment by darrelmiller** [comment]
@mohawk2 Having examples of failing descriptions would be very useful.  It would be extra useful if we could distinguish errors that the JSON SChema can detect and the ones that they can't.

**Comment by handrews** [comment]
@jberger if your question of "Are there plans for reuinification with json schema or will the split persist?" refers to the divergence between OpenAPI's Schema Object and the JSON Schema specification, that is tracked as part of #1443 and I (and other JSON Schema contributors) are actively working with the OpenAPI TSC on this.  I think everyone wants convergence, we're just working through the details.

@darrelmiller regarding YAML, JSON Schema is defined on a data model, not JSON text.  The data model matches what JSON Parsers produce, but YAML parsers will also produce something compatible with the data model unless the author uses those YAML features that 99% of people don't even know exist.  I write schemas in YAML myself whenever possible. :-)  JSON5 is a nice middle ground.

**Comment by mohawk2** [comment]
@darrelmiller @handrews Here is an early version of what I was referring to, for v2: https://github.com/jhthorsen/json-validator/pull/107

I thought it would be more useful to go a bit beyond just valid/invalid specs into semantic tests, in the same spirit as the highly impressive JSON Schema tests.

Two questions, then:
- how does it look so far in terms of direction?
- should I make the `basic.json` schemas be proper full v2 schemas, or do they provide value in the partial-schema form as they are now?

**Comment by lashchev** [comment]
Any update on official JSON schema?

I completely agree with @brandur and want to highlight one thing that we probably all here keep in mind but treat differently.

We need JSON schema not to describe the logical model of OpenAPI spec, but only to describe one of the serialization formats for these models. JSON Schema primary audience is serialization/deserialization layer/step in the process. Intellisense-kind of tools need it as well, but I guess number of such IDEs is lower.

If you write OpenAPI spec "validator" (or any validator in general): there are typically 3 phases in the "validation":
1. Media validation (check that it is a valid JSON, in general)
2. Format/structure validation (check that JSON follows some template/schema)
3. Semantical validation (check logical constraints/concepts expressed/serialized in the JSON)

So, JSON schema can help with step 2 and if you wrote/reviewed many Swagger files - you would know that absolute majority of errors you will typically find fall under format/structure validation category.

I regularly review a lot of swagger files that are created by many developers with different background and around 80% (I have stats) of my and dev team API review time is wasted on going through that "noise" - when people by mistake put wrong property names, wrong value data types, use Swagger 2 objects in OpenAPI docs and vice-a-versa (since they used to them and they can copy-paste without checking for required changes). And even after my (very careful) review we still keep finding silly issues in openapi.json and swagger.json for quite some time. Each of them can be 70-100 pages, so no wonder that manual review quality is not perfect.

Validations and reviews could've been significantly simplified (**not eliminated / fully automated**) with JSON schema.

How to make progress? I am with @KES777 here - you just go forward and fix the direction if/when needed. Waiting for PR with absolutely perfect JSON schema IMHO (and how it shows here) is not practical - results will never materialize.

Take an example of RFC process or JSON Schema project itself - they publish drafts and move forward. Will they have a release? I don't know and I don't care honestly - I care that they have very useful artifacts that I can use right now and that they improve them fairly often. 

> perfect is the enemy of done

So, I would really love seeing you guys adding [draft] JSON schema into [Schemas](https://github.com/OAI/OpenAPI-Specification/tree/master/schemas) so we all can start using it.

Do you have 2 or 10 alternatives? Just pick one (don't spend too much time picking - 15 mins max, or just roll the dice if needed ;-)), but pick one and publish. You will definitely get more/better feedback and get JSON Schema finished way faster.

**Comment by mikhailian** [comment]
@handrews, "...by the end of the year" is exactly the opposite of what @lashchev suggests.

OAS 3.0 should not have been released without a JSON schema to start with. As a developer, I am struggling to explain business people that "OAS 3.0 release" is not exactly the same as "OAS 2.0 release", and this sucks.

**Comment by handrews** [comment]
@mikhailian 
* JSON Schema for OAS 3.x: Within a few weeks.  @webron and I started this up again this week.
* Next draft ***of JSON Schema itself***, a.k.a. draft-08: by the end of the year

I was not around when OAS 3.0 was finalized and release, so I can't comment on the decision there.  I'm mainly the JSON Schema draft editor/author, and I help out here mostly on things related to that (so, supporting newer JSON Schema drafts, looking at what features we can add to JSON Schema to better support OAS, and yes, helping the TSC produce JSON Schemas for the spec).

**Comment by mikhailian** [comment]
@handrews, thanks for the commitment. 

I found your [unevaluatedProperties](https://github.com/json-schema-org/json-schema-spec/issues/556) proposal while trying to understand what's the problem with JSON Schema for OAS 3.

It looks like OAS 3.x would not have used `x-` properties if it had to produce a JSON Schema by at release time. Now, you are trying to fix OAS 3 idiosyncrasies by pushing a change to JSON Schema in a tail-wagging-the-dog move.

This is fascinating ;-)

**Comment by lashchev** [comment]
@handrews It is very excited to see JSON Schema and OAS 3.0 Schema efforts collide in such a very constructive way. I look forward to seeing the results of your very appreciated effort!

**Comment by handrews** [comment]
@mikhailian 

> It looks like OAS 3.x would not have used x- properties if it had to produce a JSON Schema at release time. Now, you are trying to fix OAS 3.x idiosyncrasies by pushing a change to JSON Schema in a tail-wagging-the-dog move.

It's hardly a tail-wagging-the-dog situation.  OpenAPI is one of the largest users of JSON Schema, which became a de-facto standard at draft-04 after the previous editors abandoned the project.  As we work through the drafts on the way towards standardization, we take advantage of our user base to see what works and what does not in real-world scenarios.

In the case of OpenAPI and `unevaluatedProperties`, this new keyword addresses a longstanding frustration of many JSON Schema users.  The difficulties of producing an OAS 3.0 spec without it helped validate that the keyword was needed, and not (as some claimed) a solution looking for a problem.  The best feedback is always real-world usage.

Similarly, looking at what differences OAS chose to have between their Schema Object and regular JSON Schema is very helpful in identifying gaps (`writeOnly`, which we shamelessly stole), confusions, or burdensome implementation requirements.  In particular, the code generation use case of OAS has highlighted that the validation vocabulary is not really designed for that use case, and can't really be shoved into it without making a mess.  That has greatly informed our planned code generation vocabulary project.

**Comment by mohawk2** [comment]
>  The best feedback is always real-world usage.

Amen!

**Comment by mikhailian** [comment]
Let's forget business interests and hierarchical relationships for a moment.

What changes should be done to OAS3.x so that the resulting JSON Schema becomes concise and generates a simple error stack in JSON Schema validators?

Would removing `x-` properties and replacing them with `additionalProperties: true` be enough?

In case you wonder what the hell this is about, I'll tell you about my real-world use case. I gave one of the alternative schemas to technically savvy non-programmers, and the feedback was that error stacks were not easy to understand and quite verbose.  Once an official JSON Schema design for OAS3.x  gets published, users won't be immediately happy.

**Comment by handrews** [comment]
OK, I am really, truly, for real working on this next week.  Stay tuned... and feel free to mock me if I don't get anything done 🙃

**Comment by handrews** [comment]
First PR on the WIP schema: #1744

**Comment by handrews** [comment]
I have more PRs that will get put up once 1744 is merged into the schema branch, because they depend on it and I don't have the right permissions to push another base branch upstream for further PRs.  I'm trying to keep the PRs reasonably sized so they are more accessible for people who would like to give feedback.

**Comment by fwouts** [comment]
I'd love to use the JSON Schema definition to validate OpenAPI 3 documents. What's the progress on this? How can I help?

**Comment by handrews** [comment]
@fwouts I've had a series of minor but distracting health problems since late november (shortly after making that comment) that have prevented me from making much progress.  I plan to post what I have as a PR (alongside another one that someone else posted) within the next week and then hopefully someone else can finish it.  I'm behind in damn near everything in life right now and just don't have the time (I haven't even been able to get the latest draft of the JSON Schema spec out, which has been stalled for the same reasons for the same amount of time).

**Comment by tnpradeep** [comment]
Any estimation when this will be released ?

**Comment by webron** [comment]
Soon. Very soon.

**Comment by MikeRalphson** [comment]
Move to close?

**Comment by webron** [comment]
Aye.

**Comment by Chaz6** [comment]
Does this mean that the schema is available now since the bug is closed?

**Comment by ahx** [comment]
@Chaz6 Yes! https://github.com/OAI/OpenAPI-Specification/tree/master/schemas/v3.0

**Comment by handrews** [comment]
@Barzi-Ahmed the schemas are published (in JSON form) at https://spec.openapis.org/

The YAML files in this repository are just the sources we use to generate the official schemas on that site.  We're trying to rearrange our repo a bit to make this less confusing.

**Comment by ralfhandl** [comment]
@Barzi-Ahmed Try using this URL:
- https://spec.openapis.org/oas/3.0/schema/2024-10-18

That is the current OpenAPI 3.0 schema.

**Comment by karenetheridge** [comment]
@Barzi-Ahmed https://spec.openapis.org/#openapi-specification-schemas
